k8s_test_harness/util/env_util.py

```diff
@@ -0,0 +1,204 @@
+# See LICENSE file for licensing details
+#
+
+"""Module containing utilities for extracting env options used in testing.
+
+This module is intended to be used with pre-built ROCK meta information
+produced by the canonical/k8s-workflows/build_rocks.yaml workflow:
+
+# TODO(aznashwan): pin exact link upon first stable release of workflows:
+https://github.com/canonical/k8s-workflows/blob/main/.github/workflows/build_rocks.yaml#L125-L133
+"""
+
+import json
+import os
+from dataclasses import dataclass
+from typing import Dict, List
+
+
+DEFAULT_BUILT_ROCKS_METADATA_ENV_VAR = "BUILT_ROCKS_METADATA"
+
+
+def must_get_env_var(var_name: str) -> str:
+    """Reads the environment variable with the given name.
+
+    :raises EnvironmentError: when the variable isn't set or is empty.
+    """
+    val = os.getenv(var_name, default='')
+    if not val:
+        raise EnvironmentError(
+            f"Environment variable '{var_name}' is not defined. "
+            f"Current environment contains: {dict(os.environ)}")
+    return val
+
+
+@dataclass
+class RockMetaInfo:
+    """Class representing metadata for a built ROCK as produced by the
+    canonical/k8s-workflows/build_rocks.yaml workflow:
+
+    # TODO(aznashwan): pin exact link upon first stable release of workflows:
+    https://github.com/canonical/k8s-workflows/blob/main/.github/workflows/build_rocks.yaml#L125-L133
+    """
+    name: str
+    version: str
+    rock_dir: str
+    arch: str
+    image: str
+    runs_on: List[str]
+    rockcraft_revision: str
+
+    @classmethod
+    def from_dict(cls, dict_val: Dict[str, str]) -> 'RockMetaInfo':
+        """ Loads fields from the given dict of the form:
+        {
+            "name": rockName,
+            "version": rockVersion,
+            "path": rockPath,
+            "arch": arch,
+            "image": image,
+            "rockcraft-revision": rockcraftRevisions[arch] || '',
+            "runs-on-labels": platformLabels[arch] || [inputs["runs-on"]]
+        }
+
+        :raises ValueError: on missing or null-valued fields.
+        """
+        kwarg_to_key_map = {
+            "name": "name",
+            "version": "version",
+            "rock_dir": "path",
+            "arch": "arch",
+            "image": "image",
+            "rockcraft_revision": "rockcraft-revision",
+            "runs_on": "runs-on-labels"
+        }
+        init_kwargs = {
+            kwarg: dict_val.get(dict_key, None)
+            for kwarg, dict_key in kwarg_to_key_map.items()}
+        unset_keys = [
+            dict_key
+            for kwarg, dict_key in kwarg_to_key_map.items()
+            if init_kwargs[kwarg] == None]
+        if unset_keys:
+            raise ValueError(
+                f"Missing ROCK build meta info fields {unset_keys} from "
+                f"the provided dict: {dict_val}")
+
+        return cls(**init_kwargs)  # pyright: ignore
+
+    @classmethod
+    def from_json_string(cls, json_str: str) -> 'RockMetaInfo':
```

Comment:
nit: you can use `typing.Self` here.

Reply:
That was [what I did initially](https://github.com/canonical/k8s-test-harness/pull/5/commits/9e524854e5927536b80231ee02100951d0efce99) but the GH runner is using Python 3.8 where it does not exist and I had to switch to using a plain string for the code to actually run. (in terms of type hinting tooling the effect should be the same)